### ISSUE TITLE
Optimize currently visible peaks

### ISSUE THREAD

**Issue by lpatiny** [cheminfo/nmrium#1674]
While we could optimise the shape of one peak individually it would be great to be able to optimise many peaks at once.

Add a button in the peak toolbar that will send:
- all the currently displayed peaks
- the {x,y} of the currently displayed zone

To https://www.npmjs.com/package/ml-spectra-fitting

And the value of the peaks should then be replaced by the optimized one.

**Comment by jobo322** [comment]
what about using ml-gsd/optimizePeaks for it, it is a wrapper of ml-spectra-fitting and makes a peak grouping that could improve the result by reducing the number of parameters to optimize when it is possible

**Comment by lpatiny** [comment]
yes you are right it should use ml-gsd/optimizePeaks

**Comment by lpatiny** [comment]
@hamed-musallam Could you add a small button in the peaks panel with mouse over giving as message: 'Optimize visible peaks'

This will call a method that will return new peaks. As input we need

- {x:[], y:[]} of the visible zone (using subarray !!!, no need to duplicate)
- visible peaks

It will return new values for the visible peaks

How to replace the previous peaks ?

**Comment by jobo322** [comment]
@lpatiny could we have not the property width for the optimization? `width` parameter is mandatory at the moment to add the missing shape object in the peak.

**Comment by jobo322** [comment]
Version 8.6.0 of nmr-processing package exports a function `xyPeakOptimization`

( data: DataXY, peaks: PeakXYWidth[],  options: [OptimizePeaksOptions](https://github.com/mljs/global-spectral-deconvolution/blob/master/src/post/optimizePeaks.ts#L10-L38)) ) => [GSDPeakOptimized](https://github.com/mljs/global-spectral-deconvolution/blob/master/src/GSDPeakOptimized.ts)[]

```js
interface PeakXYWidth {
  x: number;
  y: number;
  width: number;
}
```

**Comment by lpatiny** [comment]
xyPeakOptimization should allow the presence of a shape !

**Comment by lpatiny** [comment]
I guess that xyPeakOptimization could also recalculate width after optimization.

**Comment by jobo322** [comment]
@lpatiny the width is recalculated from ml-gsd/optimizePeaks

the shape is not mandatory but it is allowed to pass a peak with it

**Comment by lpatiny** [comment]
Ok perfect. So if we send a shape it will be taken into account (and also his fwhm) and there is a fallback to use gaussian shape and width

**Comment by hamed-musallam** [comment]
closed by #1683
